icu4j/main/classes/core/src/com/ibm/icu/impl/breakiter/CjkBreakEngine.java

```diff
@@ -14,28 +14,64 @@
 
 import java.io.IOException;
 import java.text.CharacterIterator;
+import java.util.HashSet;
 
 import com.ibm.icu.impl.Assert;
+import com.ibm.icu.impl.ICUData;
+import com.ibm.icu.lang.UCharacter;
+import com.ibm.icu.lang.UCharacterCategory;
 import com.ibm.icu.text.Normalizer;
 import com.ibm.icu.text.UnicodeSet;
+import com.ibm.icu.util.UResourceBundle;
+import com.ibm.icu.util.UResourceBundleIterator;
 
 public class CjkBreakEngine extends DictionaryBreakEngine {
     private UnicodeSet fHangulWordSet;
+    private UnicodeSet fHiraganaWordSet;
+    private UnicodeSet fNumberWordSet;
     private DictionaryMatcher fDictionary = null;
+    private HashSet<String> fParticleSet;
+    //open punctuation mask
+    private static final int U_GC_OP_MASK = (1 << UCharacterCategory.INITIAL_PUNCTUATION)
+            | (1 << UCharacterCategory.START_PUNCTUATION);
+    //close punctuation mask
+    private static final int U_GC_CP_MASK = (1 << UCharacterCategory.DASH_PUNCTUATION)
+            | (1 << UCharacterCategory.END_PUNCTUATION)
+            | (1 << UCharacterCategory.CONNECTOR_PUNCTUATION)
+            | (1 << UCharacterCategory.OTHER_PUNCTUATION)
+            | (1 << UCharacterCategory.FINAL_PUNCTUATION);
 
     public CjkBreakEngine(boolean korean) throws IOException {
         fHangulWordSet = new UnicodeSet("[\\uac00-\\ud7a3]");
         fHangulWordSet.freeze();
+        fHiraganaWordSet = new UnicodeSet("[:Hiragana:]");
```

Comment:
so, looks like both fHiraganaWordSet and fParticleSet are always used in exactly the same place to decide a break point should be skip or not. How about instead of checking a HashSet first then check a UnicodeSet, we only check one HashSet. so
1. Rename the fParticleSet to fSkipSet 
2. Remove fHiraganaWordSet from the member data
3. In the constructor, create a stack hiraganaWordSet  as how you currently create fHiraganaWordSet but then add the content of that hiraganaWordSet into fSkipSet by using https://unicode-org.github.io/icu-docs/apidoc/dev/icu4j/com/ibm/icu/text/UnicodeSet.html#iterator-- and you won't need hiraganaWordSet after the constructor.
4. skip the checking of fHiraganaWordSet since the check will be done inside fSkipSet.contains()

Reply:
Resolved